File: native/android/app/src/main/java/io/github/paulhammant/liveverifyprototype/core/VerificationMeta.kt
package io.github.paulhammant.liveverifyprototype.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.net.URI

@Serializable
data class VerificationMeta(
    val description: String? = null,
    val issuer: String? = null,
    val claimType: String? = null,
    val parentAuthorities: List<String>? = null,
    val responseTypes: Map<String, ResponseType>? = null,
    val retentionLaws: List<RetentionLaw>? = null,
    val charNormalization: String? = null,
    val ocrNormalizationRules: List<OCRNormalizationRule>? = null,
) {
    @Serializable
    data class ResponseType(
        @SerialName("class") val klass: String? = null,
        val text: String? = null,
        val link: String? = null,
    ) {
        val classification: VerificationClassification
            get() = when ((klass ?: "").lowercase()) {
                "affirming" -> VerificationClassification.AFFIRMING
                "denying" -> VerificationClassification.DENYING
                else -> VerificationClassification.UNKNOWN
            }

        val policyLink: URI?
            get() = link?.let { runCatching { URI(it) }.getOrNull() }
    }

    @Serializable
    data class RetentionLaw(
        val jurisdiction: String? = null,
        val law: String? = null,
        val link: String? = null,
        val summary: String? = null,
    )

    @Serializable
    data class OCRNormalizationRule(
        val pattern: String? = null,
        val replacement: String? = null,
        val description: String? = null,
    )
}

File: native/ios/LiveVerifyPrototype/LiveVerifyPrototype/Pipeline/VerificationMeta.swift
import Foundation

struct VerificationMeta: Decodable, Equatable {
    var description: String?
    var issuer: String?
    var claimType: String?
    var parentAuthorities: [String]?

    var responseTypes: [String: ResponseType]?

    var retentionLaws: [RetentionLaw]?

    var charNormalization: String?
    var ocrNormalizationRules: [OCRNormalizationRule]?

    struct ResponseType: Decodable, Equatable {
        var `class`: String?
        var text: String?
        var link: String?

        var classification: VerificationClassification {
            switch (`class` ?? "").lowercased() {
            case "affirming": return .affirming
            case "denying": return .denying
            default: return .unknown
            }
        }

        var policyURL: URL? {
            guard let link else { return nil }
            return URL(string: link)
        }
    }

    struct RetentionLaw: Decodable, Equatable {
        var jurisdiction: String?
        var law: String?
        var link: String?
        var summary: String?
    }

    struct OCRNormalizationRule: Decodable, Equatable {
        var pattern: String?
        var replacement: String?
        var description: String?
    }
}

